From 69e03fa6fcf63dc03b367f5f98d784982a81a668 Mon Sep 17 00:00:00 2001
From: Yiqun Lin <yqlin@apache.org>
Date: Mon, 6 Feb 2017 13:15:16 +0800
Subject: [PATCH 2609/2748] HDFS-11377. Balancer hung due to no available
 mover threads. Contributed by yunjiong zhao.

(cherry picked from commit 9cbbd1eae893b21212c9bc9e6745c6859317a667)
(cherry picked from commit 26bed5ce53b7acf888d097d9b78eea28475a62b5)

 Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Dispatcher.java

Change-Id: I2945b076c58b0f679feb89205344bc48ef460728
---
 .../hadoop/hdfs/server/balancer/Dispatcher.java    |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Dispatcher.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Dispatcher.java
index d498e3a..66eeb9a 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Dispatcher.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Dispatcher.java
@@ -85,7 +85,6 @@
   private static final long GB = 1L << 30; // 1GB
   private static final long MAX_BLOCKS_SIZE_TO_FETCH = 2 * GB;
 
-  private static final int MAX_NO_PENDING_MOVE_ITERATIONS = 5;
   /**
    * the period of time to delay the usage of a DataNode after hitting
    * errors when using it for migrating data
@@ -993,6 +992,8 @@ public void executePendingMove(final PendingMove p) {
     }
     if (moveExecutor == null) {
       LOG.warn("No mover threads available: skip moving " + p);
+      targetDn.removePendingBlock(p);
+      p.proxySource.removePendingBlock(p);
       return;
     }
 
-- 
1.7.9.5

